dotnet-authserver/src/TeacherIdentity.AuthServer/Controllers/AuthorizationController.cs

```diff
@@ -84,73 +89,49 @@ public async Task<IActionResult> Authorize()
 
             TrnRequirementType? trnRequirementType = null;
-            string? trn = null;
 
             if (userRequirements.HasFlag(UserRequirements.TrnHolder))
             {
-                var requestedTrnRequirement = request["trn_requirement"];
+                trnRequirementType = await TryGetTrnRequirementType(request);
 
-                if (requestedTrnRequirement.HasValue)
-                {
-                    if (!Enum.TryParse<TrnRequirementType>(requestedTrnRequirement?.Value as string, out var parsedTrnRequirementType))
-                    {
-                        return Forbid(
-                            authenticationSchemes: OpenIddictServerAspNetCoreDefaults.AuthenticationScheme,
-                            properties: new AuthenticationProperties(new Dictionary<string, string?>()
-                            {
-                                [OpenIddictServerAspNetCoreConstants.Properties.Error] = Errors.InvalidRequest,
-                                [OpenIddictServerAspNetCoreConstants.Properties.ErrorDescription] =
-                                    "Invalid trn_requirement specified."
-                            }));
-                    }
-
-                    trnRequirementType = parsedTrnRequirementType;
-                }
-                else
+                if (trnRequirementType is null)
                 {
-                    var client = (await _applicationManager.FindByClientIdAsync(request.ClientId!))!;
-                    trnRequirementType = client.TrnRequirementType;
+                    return Forbid(
+                        authenticationSchemes: OpenIddictServerAspNetCoreDefaults.AuthenticationScheme,
+                        properties: new AuthenticationProperties(new Dictionary<string, string?>()
+                        {
+                            [OpenIddictServerAspNetCoreConstants.Properties.Error] = Errors.InvalidRequest,
+                            [OpenIddictServerAspNetCoreConstants.Properties.ErrorDescription] =
+                                "Invalid trn_requirement specified."
+                        }));
                 }
 
                 var requestedTrnToken = request["trn_token"];
 
                 if (_configuration.GetValue("RegisterWithTrnTokenEnabled", false) && requestedTrnToken.HasValue)
                 {
-                    var trnToken = await _dbContext.TrnTokens.SingleOrDefaultAsync(t => t.TrnToken == requestedTrnToken.Value.Value as string && t.ExpiresUtc > _clock.UtcNow);
-                    if (trnToken is not null)
-                    {
-                        if (await _dbContext.Users.FirstOrDefaultAsync(u => u.Trn == trnToken.Trn) is null)
-                        {
-                            trn = trnToken.Trn;
-                        }
-                    }
+                    authStateInitData = await GetTrnTokenInitData(requestedTrnToken.Value.Value as string);
```

Comment:
Not sure what the interaction between a signed in user and trn_token is supposed to be. In this implementation the signedInUser data is simply overwritten

Reply:
Eventually we'd want to merge the signed in user's details with data from the `trn_token`. We don't have a journey for that yet, might be worth adding a `TODO` comment to the effect of `Merge this data with signed in user's data`